src/Infrastructure/FileSyncer/RsyncFileSyncer.php

```diff
@@ -46,14 +46,16 @@ public function sync(
             // Archive mode--the same as -rlptgoD (no -H), or --recursive,
             // --links, --perms, --times, --group, --owner, --devices, --specials.
             '--archive',
-            // Delete extraneous files from destination directories.
-            '--delete',
+            // Delete extraneous files from destination directories. Note: Use of
+            // --delete-during rather than merely --delete prevents "file has
+            // vanished" errors when syncing a directory with its own ancestor.
+            '--delete-during',
             // Increase verbosity.
             '--verbose',
         ];
 
-        // Prevent infinite recursion if the destination is inside the source.
-        $exclusions[] = $destination;
+        // Prevent infinite recursion if the source is inside the destination.
+        $exclusions[] = $source;
```

Comment:
This is clearly intentional, but I'd like to understand the thinking that begat it. :)

Reply:
Simply that the old code turned out not do what I thought it would do in the case in question, and new functional tests revealed it and led to the new code.